File: nodejs/cross_verify.js
const sc = require('./secure_channel')
const fs = require('fs')


function outputForNodejs(prefix) {

    // create output content

    let contentOutput = ""

    contentOutput += "# output by nodejs, verify by golang\n"
    contentOutput += "# verify need bkeypair.PrivateKey, akeypair.PublicKey, signedData (DataEnc, Hash, Sig)\n"

    for(let i = 0; i < benchCount; i++) {

        let akeypair = sc.generateECKeyPair()
        let bkeypair = sc.generateECKeyPair()

        contentOutput += bkeypair.PrivateKey + "\n"
        contentOutput += akeypair.PublicKey + "\n"

        let signedData = sc.calEncSignedData(akeypair.PrivateKey, bkeypair.PublicKey, data)

        contentOutput += signedData.DataEnc + "\n"
        contentOutput += signedData.Hash + "\n"
        contentOutput += signedData.Sig + "\n"
    }

    // write to file

    try {
        fs.writeFileSync(`./${prefix}cross_verify_output_nodejs.txt`, contentOutput)
    } catch (err) {
        console.error(`failed to write, ${err}`)
    }

	console.log("output finished")
}


function verifyNodejsOutput(prefix) {

    let inputContent = ""

    try {
        inputContent = fs.readFileSync(`${prefix}cross_verify_output_go.txt`, 'utf8')
    } catch (err) {
        console.error(`failed to read, ${err}`)
        return 
    }

    // loop input content
    let lines = inputContent.split("\n").slice(2)
    let count = parseInt(lines.length / 5)

    for(let i = 0; i < count; i++){

        let start = i * 5
        let bPrivateKey = lines[start++]
		let aPublicKey = lines[start++]
		let dataEnc = lines[start++]
		let hash = lines[start++]
		let sig = lines[start]

        let encSignedData = {
            DataEnc: dataEnc, 
            Hash: hash,
            Sig: sig
        }

        let result = sc.verifyEncSignedData(bPrivateKey, aPublicKey, encSignedData)
        
        console.log("------")
		console.log(result.Data)
		if (!result.Result) {
			console.log("failed to verify!")
		}
    }

}



let data = "Test data 1	Test data 2	Test data 3	Test data 4	Test data 5	Test data 6	Test data 7	Test data 8	Test data 9	Test data 10	Test data 11	Test data 12	Test data 13	Test data 14	Test data 15	Test data 16	Test data 17	Test data 18	Test data 19	Test data 20	Test data 21	Test data 22	Test data 23	Test data 24	Test data 25	Test data 26"

let benchCount = 10





let prefixOutput = "03.14_"
	
outputForNodejs(prefixOutput)

let prefixNodejsOutput = "../golang/secure_channel_cross_verify/03.14_"

verifyNodejsOutput(prefixNodejsOutput)

File: golang/secure_channel_cross_verify/cross_verify.go
package main

import (
	"fmt"
	"bufio"
	"os"
	sc "secure_channel"
)

func outputForNodejs(prefix string){

	// create writer

	outputFileName := prefix + "cross_verify_output_go.txt"

	outputFile, errOpenFile := os.OpenFile(outputFileName, os.O_CREATE|os.O_WRONLY, 0777)
	if errOpenFile != nil{
		fmt.Println("open file err =", errOpenFile)
		return
	}

	defer outputFile.Close()
	outputWriter := bufio.NewWriter(outputFile)
	
	// output

	data := "Test data 1	Test data 2	Test data 3	Test data 4	Test data 5	Test data 6	Test data 7	Test data 8	Test data 9	Test data 10	Test data 11	Test data 12	Test data 13	Test data 14	Test data 15	Test data 16	Test data 17	Test data 18	Test data 19	Test data 20	Test data 21	Test data 22	Test data 23	Test data 24	Test data 25	Test data 26"

	outputWriter.WriteString("# output by golang, verify by nodejs" + "\n")
	outputWriter.WriteString("# verify need bkeypair.PrivateKey, akeypair.PublicKey, signedData (DataEnc, Hash, Sig)" + "\n")

	benchCount := 10

	for i := 0; i < benchCount; i++ {
		akeypair, _ := sc.GenerateECKeyPair()
		bkeypair, _ := sc.GenerateECKeyPair()
	
		outputWriter.WriteString(bkeypair.PrivateKey + "\n")
		outputWriter.WriteString(akeypair.PublicKey + "\n")
	
		esData, _ := sc.CalEncSignedData(akeypair.PrivateKey, bkeypair.PublicKey, data)
	
		outputWriter.WriteString(esData.DataEnc + "\n")
		outputWriter.WriteString(esData.Hash + "\n")
		outputWriter.WriteString(esData.Sig + "\n")
	}

	// flush to file

	outputWriter.Flush()

	fmt.Println("output finished")
}

func verifyNodejsOutput(prefix string) {

	// create reader

	inputFileName := prefix + "cross_verify_output_nodejs.txt"

	inputFile, errOpenFile := os.OpenFile(inputFileName, os.O_RDONLY, 0777)
	if errOpenFile != nil{
		fmt.Println("open file err =", errOpenFile)
		return
	}

	defer inputFile.Close()
	
	inputReader := bufio.NewScanner(inputFile)
	
	// read and verify

	inputReader.Scan()
	inputReader.Scan() // skip comments

	for {
		if !inputReader.Scan() {
			break
		}
		// bkeypair.PrivateKey, akeypair.PublicKey, signedData (DataEnc, Hash, Sig)
		bPrivateKey := inputReader.Text()
		inputReader.Scan()
		aPublicKey := inputReader.Text()
		inputReader.Scan()
		dataEnc := inputReader.Text()
		inputReader.Scan()
		hash := inputReader.Text()
		inputReader.Scan()
		sig := inputReader.Text()

		encSignedData := &sc.EncSignedData{
			DataEnc: dataEnc,
			Hash: hash,
			Sig: sig,
		}

		data, err := sc.VerifyEncSignedData(bPrivateKey, aPublicKey, encSignedData)

		fmt.Println("------")
		fmt.Println(data)

		if err != nil {
			fmt.Println("failed to verify!")
		}
	}
}

func main() {

	prefixOutput := "03.14_"
	
	outputForNodejs(prefixOutput)


	prefixNodejsOutput := "../../nodejs/03.14_"

	verifyNodejsOutput(prefixNodejsOutput)
}
